### PR TITLE
Il2cpp bee

### PR DESCRIPTION
**Pull request by lucasmeijer**
## Tundra changes
* It was very easy for tundra to pick up old .dag.json files.  This causes frequent buildfailures in graphs that use nodes that have targetDirectories,  where something goes wrong in between generating the .json and tundra baking it.   We've fixed this by making it explicit when invoking tundra that you want a specific json to be baked.  If you tell tundra to do so, tundra will always do it.   If you do not tell tundra to do so, tundra will never attempt to bake a leftover json file from the past.
* Change tundra behaviour to eagerly enqueue nodes to avoid the current-situation-on-trunk where the progress indicator has a slowly increasing total amount of nodes.
* Add the concept of statsignatures to tundra. Previously when in bee you do .FileExists(), we turned that into a FileSignature. This causes a graph rebuild anytime that file changes. This is unnecissery since we only need to rebuild the graph if the file stops (or starts) existing. 
* rewrite how we files that were created in dynamic output directories to not have a memory limit.  IL2CPP was hitting this memory limit.

### CODE REVIEW

**Comment by mostynb** [comment]
Meta comment: I think this would be easier to review as four(?) separate PRs.

Looking anyway...

File: src/AllBuiltNodes.cpp [hunk 1]
```diff
@@ -156,6 +155,7 @@ bool SaveAllBuiltNodes(Driver *self)
             case NodeBuildResult::kUpToDate:
             case NodeBuildResult::kRanSuccesfully:
             case NodeBuildResult::kRanSuccessButDependeesRequireFrontendRerun:
+            case NodeBuildResult::kUpToDataButDependeesRequireFrontendRerun:
```
Comment (mostynb) [inline src/AllBuiltNodes.cpp]: Should this be "up to date" instead of "up to data" ?

File: src/BuildLoop.cpp [hunk 2]
```diff
@@ -74,45 +75,24 @@ static void LogFirstTimeEnqueue(MemAllocLinear* scratch, RuntimeNode* enqueuedNo
...
 }
 
-static int EnqueueNodeListReversedWithoutWakingAwaitersOnlyWhenNotEnqueuedBefore(BuildQueue* queue, MemAllocLinear* scratch, const FrozenArray<int32_t>& nodesToEnqueue, RuntimeNode* enqueingNode)
+static int EnqueueNodeListWithoutWakingAwaiters(BuildQueue* queue, MemAllocLinear* scratch, const FrozenArray<int32_t>& nodesToEnqueue, RuntimeNode* enqueingNode)
 {
-    int enqueue_count = 0;
+    int placed_on_workstack_count = 0;
 
     //we enqueue dependency lists in reverse. because our semantics say that the most urgent dependencies are in the list first, it's important that they
     //end up on the top of the workstack, so they'll be processed first.
```
Comment (mostynb) [inline src/BuildLoop.cpp]: Does this comment still make sense here? It looks like we process nodesToEnqueue in "forward" order now.

File: src/BuildLoop.cpp [hunk 3]
```diff
@@ -139,9 +119,67 @@ static bool AllDependenciesAreSuccesful(BuildQueue *queue, RuntimeNode *runtime_
...
+    //know they cannot yet immediately be acted upon. They will be put on the workstack when their dependencies finish.
+    int placed_on_workstack_count = 0;
+    if (AllDependenciesAreFinished(queue,runtime_node) || IsNodeCacheableByLeafInputsAndCachingEnabled(queue,runtime_node))
+    {
+        if (AddNodeToWorkStackIfNotAlreadyPresent(queue, runtime_node))
+            placed_on_workstack_count++;
+    } else {
+        //ok, so in this case the queued node is not immediately actionable. so we don't put it on the workstack, but instead queue up all our tobuild dependencies,
+        //so that our node becomes immediately actionable in the future. We don't blindly always do this, because in the case where this node is leaf input cacheable
+        //we might get a cache hit, and it won't be necissery at all to build any of the dependencies.
```
Comment (mostynb) [inline src/BuildLoop.cpp]: nitpick: s/necissery/necessary/ (I'm bad at spelling that word too)

File: src/AllBuiltNodes.cpp (hunk 1 above)
Comment (bitter) [reply src/AllBuiltNodes.cpp]: fixed 8882e09

File: src/BuildLoop.cpp (hunk 3 above)
Comment (bitter) [reply src/BuildLoop.cpp]: fixed 8882e09

File: src/BuildLoop.cpp (hunk 2 above)
Comment (bitter) [reply src/BuildLoop.cpp]: fixed 8882e09

**Review by mostynb — APPROVED** [review]

**Review by bitter — APPROVED** [review]

**Pull request merged**